### PR TITLE
Ldap backend update

### PR DESCRIPTION
**Pull request by goestreicher**
As discussed in issue #3358 here is the port of the various modifications done on the LDAP backend. I've tested locally, but not with Travis, so we'll see what comes out :)

### CODE REVIEW

**Comment by goestreicher** [comment]
Hey,

Any updates on this?

Cheers,
Grégory

**Comment by Habbie** [comment]
Reviewing this. Can you fix the doc conflict?

File: modules/ldapbackend/Makefile.am [hunk 1]
```diff
@@ -10,4 +10,4 @@ libldapbackend_la_SOURCES = \
 	ldapauthenticator.hh ldapauthenticator_p.hh ldapauthenticator.cc
 
 libldapbackend_la_LDFLAGS = -module -avoid-version
-libldapbackend_la_LIBADD = $(LDAP_LIBS)
+libldapbackend_la_LIBADD = $(LDAP_LIBS) -lkrb5
```
Comment (Habbie) [inline modules/ldapbackend/Makefile.am]: could you make this a variable filled in from the m4?

File: modules/ldapbackend/powerldap.cc [hunk 2]
```diff
@@ -156,11 +156,33 @@ void PowerLDAP::simpleBind( const string& ldapbinddn, const string& ldapsecret )
 int PowerLDAP::search( const string& base, int scope, const string& filter, const char** attr )
 {
         int msgid, rc;
-
-        if( ( rc = ldap_search_ext( d_ld, base.c_str(), scope, filter.c_str(), const_cast<char**> (attr), 0, NULL, NULL, NULL, LDAP_NO_LIMIT, &msgid ) ) != LDAP_SUCCESS )
-        {
-        	throw LDAPException( "Starting LDAP search: " + getError( rc ) );
-        }
+	bool finished = false;
```
Comment (Habbie) [inline modules/ldapbackend/powerldap.cc]: What is the purpose of the prefetch?

File: modules/ldapbackend/powerldap.cc [hunk 3]
```diff
@@ -156,11 +156,33 @@ void PowerLDAP::simpleBind( const string& ldapbinddn, const string& ldapsecret )
...
+
+	while ( !m_results.empty() ) {
+		ldap_msgfree( m_results.front() );
+		m_results.pop_front();
```
Comment (Habbie) [inline modules/ldapbackend/powerldap.cc]: `pop_front` is expensive. I suggest iterating, or using `pop_back`.

File: modules/ldapbackend/powerldap.cc [hunk 4]
```diff
@@ -195,17 +217,11 @@ bool PowerLDAP::getSearchEntry( int msgid, sentry_t& entry, bool dn, int timeout
...
+	if ( m_results.empty() )
+		return false;
 
-        if( i != LDAP_RES_SEARCH_ENTRY )
-        {
-        	ldap_msgfree( result );
-        	throw LDAPException( "Search returned an unexpected result" );
-        }
+	result = m_results.front();
+	m_results.pop_front();
```
Comment (Habbie) [inline modules/ldapbackend/powerldap.cc]: Again, `pop_front` is expensive. How about just keeping an index and doing cleanup at the end (or, as you do now, at the beginning of the next query)?

File: modules/ldapbackend/powerldap.cc [hunk 5]
```diff
@@ -156,11 +156,33 @@ void PowerLDAP::simpleBind( const string& ldapbinddn, const string& ldapsecret )
...
+		m_results.pop_front();
+	}
+
+        rc = ldap_search_ext( d_ld, base.c_str(), scope, filter.c_str(), const_cast<char**> (attr), 0, NULL, NULL, NULL, LDAP_NO_LIMIT, &msgid );
+        if ( rc != LDAP_SUCCESS )
+		throw LDAPException( "Starting LDAP search: " + getError( rc ) );
+
+	while ( !finished ) {
+		LDAPMessage *result = NULL;
+		int i = waitResult( msgid, 5, &result );
```
Comment (Habbie) [inline modules/ldapbackend/powerldap.cc]: You accidentally hardcoded the timeout.

**Comment by Habbie** [comment]
In the commit text for 22ea323, replace 'cached' by 'fetched all at once' or similar.

File: modules/ldapbackend/ldapbackend.cc [hunk 6]
```diff
@@ -216,6 +245,14 @@ void LdapBackend::lookup( const QType &qtype, const DNSName &qname, DNSPacket *d
         	L << Logger::Warning << m_myname << " Unable to search LDAP directory: " << lt.what() << endl;
         	throw( DBException( "LDAP server timeout" ) );
         }
+        catch( LDAPNoConnection &lnc )
+        {
+		L << Logger::Warning << m_myname << " Connection to LDAP lost, trying to reconnect" << endl;
+		if ( reconnect() )
+			this->lookup( qtype, qname, dnspkt, zoneid );
+		else
+			throw PDNSException( "Failed to reconnect to LDAP server" );
```
Comment (Habbie) [inline modules/ldapbackend/ldapbackend.cc]: @pieterlexis @mind04 is this the right exception?

File: modules/ldapbackend/ldapbackend.cc [hunk 7]
```diff
@@ -85,6 +86,7 @@ LdapBackend::LdapBackend( const string &suffix )
 
         	string bindmethod = getArg( "bindmethod" );
         	if ( bindmethod == "gssapi" ) {
+                        setenv( "KRB5CCNAME", getArg( "krb5-ccache" ).c_str(), 1 );
```
Comment (Habbie) [inline modules/ldapbackend/ldapbackend.cc]: line looks misaligned

File: docs/markdown/authoritative/backend-ldap.md [hunk 8]
```diff
@@ -116,6 +116,48 @@ ldap-host (default "ldap://127.0.0.1:389/") : The values assigned to this parame
...
 (default "(:target:)" ) : LDAP filter for limiting IP or name lookups, e.g. (&(:target:)(active=yes)) for returning only entries whose attribute "active" is set to "yes".
 
+# Master Mode
+
+Schema update
+-------------
+
+First off adding master support to the LDAP backend needs
+a schema update. This is required as some metadata must
+be stored by PowerDNS, such as the last successfull transfer
```
Comment (Habbie) [inline docs/markdown/authoritative/backend-ldap.md]: successful

File: docs/markdown/authoritative/backend-ldap.md [hunk 9]
```diff
@@ -116,6 +116,48 @@ ldap-host (default "ldap://127.0.0.1:389/") : The values assigned to this parame
...
+to slaves. The new schema is available in
+schema/pdns-domaininfo.schema.
+
+Once the schema is loaded the zones for which you want to
+be a master must be modified. The dn of the SOA record
+*must* have the object class PdnsDomain, and thus the
```
Comment (Habbie) [inline docs/markdown/authoritative/backend-ldap.md]: Please use `` liberally to format console/editor text.

File: docs/markdown/authoritative/backend-ldap.md [hunk 10]
```diff
@@ -116,6 +116,48 @@ ldap-host (default "ldap://127.0.0.1:389/") : The values assigned to this parame
...
+backend. Furthermore the PdnsDomainType must be equal
+to 'master' (lower case).
+
+Example
+-------
+
+Here is an example LDIF of a zone that's ready for master
+operation (assuming the 'tree' style):
+
+dn: dc=example,dc=com,ou=dns,dc=mycompany,dc=com
```
Comment (Habbie) [inline docs/markdown/authoritative/backend-ldap.md]: put ``` around longer blocks of code/config/..

File: modules/ldapbackend/exceptions.hh [hunk 11]
```diff
@@ -43,3 +43,5 @@ class LDAPNoConnection : public LDAPException
 };
 
 #endif // LDAPEXCEPTIONS_HH
+
+// vim: ts=2 sw=2 sts=2 et
```
Comment (Habbie) [inline modules/ldapbackend/exceptions.hh]: please remove the vim modelines

**Comment by goestreicher** [comment]
I've updated the fork, Travis is happy, you can check that I didn't miss anything :)

**Pull request merged**